model_specific/SM/standard_model.cpp

```diff
@@ -941,91 +941,55 @@ double Standard_model::calculate_delta_alpha_s(double alphaS) const
-      piwwtMW_corrected =
-         Weinberg_angle::replace_mtop_in_self_energy_w(
-            self_energy_w_at_mw, mw_pole, se_data);
-      piwwt0_corrected =
-         Weinberg_angle::replace_mtop_in_self_energy_w(piwwt0, 0.,
-            se_data);
-   }
 
    const auto get_mh_pole = [&] () {
-      double mh_pole = get_physical().Mhh;
-      if (mh_pole == 0) {
-         mh_pole = input.get(Physical_input::mh_pole);
-      }
+      double mh_pole = this->get_physical().Mhh;
       if (mh_pole == 0) {
          mh_pole = Electroweak_constants::MH;
       }
       return mh_pole;
    };
 
-   Weinberg_angle::Data data;
-   data.scale               = scale;
-   data.alpha_em_drbar      = alpha_em_drbar;
-   data.alpha_s_mz          = qedqcd.displayAlphaSInput();
-   data.fermi_contant       = qedqcd.displayFermiConstant();
-   data.self_energy_z_at_mz = pizztMZ_corrected;
-   data.self_energy_w_at_mw = piwwtMW_corrected;
-   data.self_energy_w_at_0  = piwwt0_corrected;
-   data.mw_pole             = mw_pole;
-   data.mz_pole             = mz_pole;
-   data.mt_pole             = mt_pole;
-   data.mh_drbar            = mh_drbar;
-   data.mh_pole             = get_mh_pole();
-   data.gY                  = gY;
-   data.g2                  = g2;
-   data.g3                  = g3;
-
-   Weinberg_angle weinberg;
-   weinberg.disable_susy_contributions();
-   weinberg.set_number_of_loops(threshold_corrections.sin_theta_w);
-   weinberg.set_data(data);
-
-   const int error = weinberg.calculate();
-
-   theta_w = ArcSin(weinberg.get_sin_theta());
-
-   if (error)
-      problems.flag_no_sinThetaW_convergence();
-   else
-      problems.unflag_no_sinThetaW_convergence();
+   const double mt_pole = qedqcd.displayPoleMt();
+   const double alpha_s_mz = qedqcd.displayAlphaSInput();
+   const double dalpha_s_5_had = Electroweak_constants::delta_alpha_s_5_had;
+   const double mh_pole = get_mh_pole();
+
+   const auto sm_mw = flexiblesusy::sm_mw::calculate_mw_pole_SM_fit_MSbar(
+      this->get_physical().Mhh, mt_pole, alpha_s_mz, dalpha_s_5_had);
 
-   return std::make_pair(theta_w, weinberg.get_mw_pole());
+   qedqcd.setPoleMW(sm_mw.first);
+
+   weinberg_angle::Weinberg_angle::Sm_parameters sm_pars;
+   sm_pars.fermi_constant = qedqcd.displayFermiConstant();
+   sm_pars.mz_pole = qedqcd.displayPoleMZ();
+   sm_pars.mt_pole = mt_pole;
+   sm_pars.mh_pole = mh_pole;
+   sm_pars.alpha_s = calculate_alpha_s_SM5_at(qedqcd, qedqcd.displayPoleMt());
+   sm_pars.alpha_s_mz = qedqcd.displayAlphaSInput();
+   sm_pars.dalpha_s_5_had = dalpha_s_5_had;
+   sm_pars.mw_pole = sm_mw.first;
+
+   const int number_of_iterations =
+       std::max(20, static_cast<int>(std::abs(-log10(this->get_precision()) * 10)
+          ));
+
+   using namespace weinberg_angle;
+   Weinberg_angle weinberg(this, sm_pars);
+   weinberg.set_number_of_loops(this->get_threshold_corrections().sin_theta_w);
+   weinberg.set_number_of_iterations(number_of_iterations);
+
+   double theta_w;
```

Comment:
If an exception is thrown, then this variable contains garbage. Please initialize it so some reasonable value, maybe to the old value or the one from `Electroweak_constants`?

Reply:
Done in 97eaea3f50f6868be37489878f28d48f1e509a90.